pennylane/operation.py

```diff
@@ -300,48 +394,92 @@ def hash(self):
-        >>> array([[ 0.96891242+0.j, -0.24740396+0.j],
-                   [ 0.24740396+0.j,  0.96891242+0.j]])
+        **Example**
 
-        Returns:
-            array: matrix representation
+        >>> qml.CNOT.compute_matrix()
+        [[1, 0, 0, 0],
+        [0, 1, 0, 0],
+        [0, 0, 0, 1],
+        [0, 0, 1, 0]]
+
+        The matrix representation may depend on parameters or hyperparameters:
+
+        >>> qml.Rot.compute_matrix(0.1, 0.2, 0.3)
+        [[ 0.97517033-0.19767681j -0.09933467+0.00996671j]
+         [ 0.09933467+0.00996671j  0.97517033+0.19767681j]]
+
+        If parameters are tensors, a tensor of the same type is returned:
+
+        >>> res = qml.Rot.compute_matrix(torch.tensor(0.1), torch.tensor(0.2), torch.tensor(0.3))
+        >>> type(res)
+        <class 'torch.Tensor'>
         """
         raise NotImplementedError
 
-    @property
-    def matrix(self):
-        r"""Matrix representation of an instantiated operator
-        in the computational basis.
+    def matrix(self, wire_order=None):
+        r"""Matrix representation of this operator in the computational basis.
 
-        **Example:**
+        If ``wire_order`` is provided, the
+        numerical representation considers the position of the
+        operator's wires in the global wire order.
+        Otherwise, the wire order defaults to ``self.wires``.
+
+        If the matrix depends on trainable parameters, the result
+        will be cast in the same autodifferentiation framework as the parameters.
 
-        >>> U = qml.RY(0.5, wires=1)
-        >>> U.matrix
-        >>> array([[ 0.96891242+0.j, -0.24740396+0.j],
-                   [ 0.24740396+0.j,  0.96891242+0.j]])
+        .. note::
+            By default, this method calls the static method ``compute_matrix``,
+            which is used by subclasses to define the actual matrix representation.
+
+        A ``NotImplementedError`` is raised if the matrix representation has not been defined.
+
+        Args:
+            wire_order (Iterable): global wire order, must contain all wire labels from this operator's wires
 
         Returns:
-            array: matrix representation
+            tensor_like: matrix representation
+
+        **Example**
+
+        >>> U = qml.PauliX(wires="b")
+        >>> U.matrix()
+        [[0 1]
+         [1 0]]
+        >>> U.matrix(wire_order=["a", "b"])
+        [[0 1 0 0]
+         [1 0 0 0]
+         [0 0 0 1]
+         [0 0 1 0]]
+        >>> qml.RY(tf.Variable(0.5), wires="b").matrix()
+        tf.Tensor([[ 0.9689124  -0.24740396]
+                   [ 0.24740396  0.9689124 ]], shape=(2, 2), dtype=float32)
+
         """
-        return self._matrix(*self.parameters)
+        canonical_matrix = self.compute_matrix(*self.parameters, **self.hyperparameters)
+
+        if wire_order is None or self.wires == Wires(wire_order):
+            return canonical_matrix
+
+        return expand_matrix(canonical_matrix, wires=self.wires, wire_order=wire_order)
```

Comment:
I thought more about `qml.matrix(op)`, and I am not 100% sure we should do it. Users should be able to get info from an operator without knowing some other methods defined in other modules! I think this would be quite important...

Reply:
I think my motivation here is 99% for the documentation. If we have `qml.matrix(op)` there is _one_ main docstring with canonical examples, description, gotchas, etc.

In this approach however, there are many different `SomeOperator.matrix` that show up in the docs and google results that need to be maintained, found by the user, etc:

![image](https://user-images.githubusercontent.com/2959003/146742959-16bf9a27-c12a-49df-890e-fbe253663da2.png)

Reply:
I often think of this as well:
https://github.com/numpy/numpy/issues/7452

> If we could go back in time then I'd probably argue for not having these methods on ndarray at all, and using functions exclusively. In python, methods are a closed set of operations that have to be specified at class creation time, can't be extended, and can be overridden by subclasses; functions are an open set of operations that can be added to at any time, namespaced, and can't be overridden by subclasses (without some special work anyway). In the ndarray case, it's turned out that method subclassing is ineffective, and that there are many many operations with new ones added all the time, that these can't be written in terms of a small finite set of core operations, and thus that we need to solve the problem of overriding functions for new types anyway. So this all argues for using functions exclusively (sorry if you find it less readable). Too bad we don't have a time machine :-)

Reply:
this is not a comment on this PR, just a general thought!